libfastpm/powerspectrum.c

```diff
@@ -415,3 +415,106 @@ fastpm_powerspectrum_init_from_string(FastPMPowerSpectrum * ps, const char * str
     }
     return 0;
 }
+
+
+/* FIMXE: Probably shouldn't have belwo in this file, but it's similar to above, so perhaps can merge? */
```

Comment:
There is also a trick that you can overlay one type from another

typedef struct {
   FastPMFuncK base;
  ...
  ...
} PowerSpectrum;
PowerSpectrum ps[1];

Then you can use ((FastPMFuncK*)ps)->k, ((FastPMFuncK*)ps)->p; or ps->base.k ps->base->p.
C compiler guarantee this works.

I think you can avoid the code redundancy this way. e.g. the power spectrum functions will just do a cast and call the FastPMFuncK routines.

Reply:
Would this mean we could also remove the duplicate functions? Because PowerSpectrum has some extra attributes, like 'edges', that need to be computed. I think this would make it hard to generalise functions unless we use lots of ifs to check for attributes, but that might be ugly?

Reply:
I pushed a commit.